README.md

```diff
@@ -90,35 +86,58 @@ The above error message indicates that the Statement's Verb property has an inco
 
 ### Validation on Pattern
 
-Each Pattern is essentially a regular expression on Statement Templates, which can be composed from other Patterns. Internally, after compilation with `profile->fsms`, each Pattern is implemented as a finite state machine (FSM), which is mathematically equivalent to regular expressions.
+Each Pattern is essentially a regular expression on Statement Templates, which can be composed from other Patterns. Internally, after compilation with `compile-profiles->fsms`, each Pattern is returned as a map of the following:
+```clojure
+  {"profile-id" {"pattern-id" {:id "pattern-id"
+                               :dfa {...}
+                               :nfa {...}}}}
+```
+with `:dfa` and `:nfa` being two different FSMs:
+- `:dfa` is a (mostly: see below) deterministic, minimized FSM used for efficient matching of Statements against a Pattern.
+- `:nfa` is a non-deterministic NFA with pattern metadata associated with each of its states. This is an optional value; if present, it is used to reconstruct the path from the primary pattern to the template when constructing match failure data.
 
-There are five different types of Patterns, based on which of the five following properties they have. The `sequence` and `alternates` properties are arrays of identifiers, while `zeroOrMore`, `oneOrMore` and `optional` give a map of a single identifier. The following description are taken from the [Profile section of the Profile spec](https://github.com/adlnet/xapi-profiles/blob/master/xapi-profiles-structure.md#patterns):
+(NOTE: Unlike "true" DFAs, `:dfa` allows for some level of non-determinism, since a Statement may match against multiple Templates.)
```

Comment:
I'm curious, in practice what's the behavior in this case with ambiguous templates?

Reply:
(For example) In goes: `#{{:state 0 :accepted? false}}`
And out goes `#{{:state 1 :accepted? false} {:state 2 :accepted? true}}`

Reply:
right, but it would happily then advance either one on the next run? both?

Reply:
Depends on the FSM :) But yeah the matching will then happen from both states 1 and 2.

Reply:
or something undefined from the platform level like `(first #{<a set>})`

Reply:
Haha so I can probably make a malformed profile that keeps splitting states on these indefinitely in `xapipe` ;)

Reply:
but we could always make a malformed profile, so this is all good in my book

Reply:
Pretty sure the profile validation rules will prevent you from doing so. But if you turn `:validate-profiles?` off then it's actually pretty easy. How exactly it can be done I will leave as an exercise to the reader.